__--outline--__

In [1]:
# Title
# from The National Renewable Energy Laboratory (NREL) Rooftop Energy Potential of Low Income Communities in America (REPLICA)

# Abstract
# ...analysis of homes qualified for rooftop solar deployment; supplementary spatial data analysi in Tableau

# Scope (notebook)
# CSV->SQLite3 query->Pandas subset->CSV(s)

# Setup
# import

# Obtain
# CSV; columns 

# Scrub 
create db; verify table(s)

# Explore
# solar suitable
# owner occupied
# subsidies available

# Model
#

# References
#

## Setup

In [2]:
import os
import numpy as np
import pandas as pd
import sqlite3

## Obtain

In [3]:
# listing files in the current directory
os.listdir()

['.git', '.gitignore', '.ipynb_checkpoints', 'data', 'README.md', 'sol.ipynb']

In [4]:
# previewing data directory
os.listdir('data/')

['.ipynb_checkpoints',
 'archive',
 'rep.sqlite',
 'SEEDSII REPLICA Documentation.pdf',
 'seeds_ii_replica.csv']

Our dataset originates from:\ Mooney, Meghan; Sigrin, Ben (2018): Rooftop Energy Potential of Low Income Communities in America REPLICA. National Renewable Energy Laboratory. https://data.nrel.gov/submissions/81

The CSV flat file comprises a variety of sources. We will instantiate the data as a Pandas dataframe and view the first rows.

In [5]:
# 'https://data.nrel.gov/system/files/81/seeds_ii_replica%20%281%29.csv'
data_df = pd.read_csv('data/seeds_ii_replica.csv', nrows=10)

# viewing first rows
data_df.head()

geoid         gisjoin  state_fips  state_name state_abbr  \
0  6001400100  G0600010400100           6  California         CA   
1  6001408100  G0600010408100           6  California         CA   
2  6001409000  G0600010409000           6  California         CA   
3  6001409900  G0600010409900           6  California         CA   
4  6001422000  G0600010422000           6  California         CA   

      county_name  county_fips  tract_fips   area_km2  centroid_x  ...  \
0  Alameda County            1      400100   6.894336 -122.231947  ...   
1  Alameda County            1      408100   8.935563 -122.157212  ...   
2  Alameda County            1      409000  16.927907 -122.212192  ...   
3  Alameda County            1      409900   4.332550 -122.135826  ...   
4  Alameda County            1      422000   4.727481 -122.302847  ...   

   climate_zone_description  moisture_regime          locale  total_units  \
0                      Warm                C    City - Large          NaN   
1                      Warm                C    City - Large          NaN   
2                      Warm                C    City - Large          NaN   
3                      Warm                C    City - Large          NaN   
4                      Warm                C  City - Midsize          NaN   

   active_subsidies  avg_months_tenancy  fmr_2br  occ_rate  pct_eli_hh  \
0               NaN                 NaN      NaN       NaN         NaN   
1               NaN                 NaN      NaN       NaN         NaN   
2               NaN                 NaN      NaN       NaN         NaN   
3               NaN                 NaN      NaN       NaN         NaN   
4               NaN                 NaN      NaN       NaN         NaN   

   lihtc_qualified  
0            False  
1            False  
2             True  
3            False  
4             True  

[5 rows x 220 columns]

In [6]:
# viewing all attribute headers
for col in data_df.columns:
    print(col)

geoid
gisjoin
state_fips
state_name
state_abbr
county_name
county_fips
tract_fips
area_km2
centroid_x
very_low_mf_own_hh
very_low_mf_rent_hh
very_low_sf_own_hh
very_low_sf_rent_hh
low_mf_own_hh
low_mf_rent_hh
low_sf_own_hh
low_sf_rent_hh
mod_mf_own_hh
mod_mf_rent_hh
mod_sf_own_hh
mod_sf_rent_hh
mid_mf_own_hh
mid_mf_rent_hh
mid_sf_own_hh
mid_sf_rent_hh
high_mf_own_hh
high_mf_rent_hh
high_sf_own_hh
high_sf_rent_hh
very_low_mf_own_bldg_cnt
very_low_mf_rent_bldg_cnt
very_low_sf_own_bldg_cnt
very_low_sf_rent_bldg_cnt
low_mf_own_bldg_cnt
low_mf_rent_bldg_cnt
low_sf_own_bldg_cnt
low_sf_rent_bldg_cnt
mod_mf_own_bldg_cnt
mod_mf_rent_bldg_cnt
mod_sf_own_bldg_cnt
mod_sf_rent_bldg_cnt
mid_mf_own_bldg_cnt
mid_mf_rent_bldg_cnt
mid_sf_own_bldg_cnt
mid_sf_rent_bldg_cnt
high_mf_own_bldg_cnt
high_mf_rent_bldg_cnt
high_sf_own_bldg_cnt
high_sf_rent_bldg_cnt
very_low_mf_own_devp_cnt
very_low_mf_rent_devp_cnt
very_low_sf_own_devp_cnt
very_low_sf_rent_devp_cnt
low_mf_own_devp_cnt
low_mf_rent_devp_cnt
low_sf_

## Scrub

### Create SQLite3 Database

In [8]:
# create a new database file "conn":
conn = sqlite3.connect("data/rep.sqlite")

In [9]:
# load the CSV in chunks:
for row in pd.read_csv(
    "data/seeds_ii_replica.csv", chunksize=1000):
    # append each row to a new table "seed" in the database:
    row.to_sql("seed", conn, if_exists="append")

    # index on the 'geoid' column:
conn.execute("CREATE INDEX geoid ON seed(geoid)") 
conn.close()

### connect to sqlite database

In [10]:
# creating a connection to the database
conn = sqlite3.connect("data/rep.sqlite")

# instantiating a cursor
cur = conn.cursor()

### listing the sqlite table names

In [11]:
# getting names of all tables in the sql database
sql_tables = conn.execute("SELECT name FROM sqlite_master WHERE type='table';")

# assign table names to a variable as a list
table_list = list(map(lambda x: x[0], sql_tables.fetchall()))

# viewing/verifying table names
table_list

['seed']

Correctly, `seed` is shown as the only table in the database.

## Explore

We can review our data description to select an area of investigation

### Custom Query

Let's find all of `data_df`'s column names for owner occupied properties that are listed as suitable buildings (those including the string "own_bldg_cnt").

In [12]:
# column names for owner-occupied, solar-suitable buildings
owner_suitable = [col for col in data_df.columns if "own_bldg_cnt" in col]
owner_suitable

['very_low_mf_own_bldg_cnt',
 'very_low_sf_own_bldg_cnt',
 'low_mf_own_bldg_cnt',
 'low_sf_own_bldg_cnt',
 'mod_mf_own_bldg_cnt',
 'mod_sf_own_bldg_cnt',
 'mid_mf_own_bldg_cnt',
 'mid_sf_own_bldg_cnt',
 'high_mf_own_bldg_cnt',
 'high_sf_own_bldg_cnt']

The columns listed in `owner_suitable` report the total annual solar power generation potential (MWh) for solar-suitable buildings.

Let's query these data for tracts with active subsidies.

In [28]:
# creating a connection to the database
conn = sqlite3.connect("data/rep.sqlite")

# instantiating a cursor
cur = conn.cursor()

In [29]:
# querying attributes
cur.execute("""SELECT geoid, gisjoin, 
                state_abbr, county_name, pop_total, 
                active_subsidies, avg_monthly_bill_dlrs, dlrs_kwh,
                very_low_mf_own_bldg_cnt, very_low_sf_own_bldg_cnt, 
                low_mf_own_bldg_cnt, low_sf_own_bldg_cnt, 
                mod_mf_own_bldg_cnt, mod_sf_own_bldg_cnt, 
                mid_mf_own_bldg_cnt, mid_sf_own_bldg_cnt, 
                high_mf_own_bldg_cnt, high_sf_own_bldg_cnt
               FROM seed
               WHERE active_subsidies > 0
            """);

df = pd.DataFrame(cur.fetchall())
df.columns = [x[0] for x in cur.description]

# exiting database connection    
conn.close()

# previewing the dataframe
df

geoid         gisjoin state_abbr        county_name  pop_total  \
0      41017000500  G4100170000500         OR   Deschutes County       5847   
1      41017000800  G4100170000800         OR   Deschutes County       6909   
2      41017000900  G4100170000900         OR   Deschutes County       7495   
3      53033025006  G5300330025006         WA        King County       6008   
4      53033027701  G5300330027701         WA        King County       5565   
...            ...             ...        ...                ...        ...   
25690  45019002802  G4500190002802         SC  Charleston County       2726   
25691  27139080302  G2701390080302         MN       Scott County      15090   
25692  34009021300  G3400090021300         NJ    Cape May County       3957   
25693  34023008900  G3400230008900         NJ   Middlesex County       4004   
25694  34025805500  G3400250805500         NJ    Monmouth County       3702   

       active_subsidies  avg_monthly_bill_dlrs  dlrs_kwh  \
0                   1.0             118.688180  0.088055   
1                   4.0             101.412973  0.110601   
2                  13.0             101.412973  0.110601   
3                   2.0              96.376683  0.111160   
4                   7.0              96.376683  0.111160   
...                 ...                    ...       ...   
25690               1.0             163.098857  0.145642   
25691               3.0              83.731445  0.131830   
25692               1.0             123.070925  0.173899   
25693               1.0              97.396944  0.164645   
25694               2.0             107.111038  0.133575   

       very_low_mf_own_bldg_cnt  very_low_sf_own_bldg_cnt  \
0                        0.2681                   85.9520   
1                        0.3102                  165.8723   
2                        2.7129                  341.7480   
3                        0.8993                   21.1499   
4                        3.0984                   69.5288   
...                         ...                       ...   
25690                    0.6422                   21.5491   
25691                    5.3075                   86.5339   
25692                    5.3800                   90.2319   
25693                    3.2189                    6.0137   
25694                    3.2281                  108.2309   

       low_mf_own_bldg_cnt  low_sf_own_bldg_cnt  mod_mf_own_bldg_cnt  \
0                   1.3405              67.1244               1.8767   
1                   0.2979             104.7614               2.6930   
2                   2.1251             205.4382               3.9131   
3                   1.3617              28.1999               1.3257   
4                   5.6797             100.4304               4.0178   
...                    ...                  ...                  ...   
25690               0.6421              14.9186               1.1227   
25691              10.6150             158.2085              13.9900   
25692               8.2768             145.4222              12.5307   
25693               7.5152              12.3071               5.2842   
25694               3.1532              85.4010               3.6254   

       mod_sf_own_bldg_cnt  mid_mf_own_bldg_cnt  mid_sf_own_bldg_cnt  \
0                 130.1559               1.6086             172.7225   
1                 239.5933               2.9068             278.3938   
2                 267.7513               2.5207             294.0396   
3                  35.2499               4.0874             100.2664   
4                 117.4264               4.2618             149.8731   
...                    ...                  ...                  ...   
25690              53.8727               3.6657              96.1420   
25691             270.9649              38.4172             578.6411   
25692             254.9269              11.8824             176.0835   
25693               8.140

In [30]:
# # exiting database connection    
conn.close()

In [16]:
# viewing info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25695 entries, 0 to 25694
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   geoid                     25695 non-null  int64  
 1   gisjoin                   25695 non-null  object 
 2   centroid_x                25695 non-null  float64
 3   state_abbr                25695 non-null  object 
 4   county_name               25695 non-null  object 
 5   pop_total                 25695 non-null  int64  
 6   active_subsidies          25695 non-null  float64
 7   avg_monthly_bill_dlrs     25695 non-null  float64
 8   dlrs_kwh                  25695 non-null  float64
 9   very_low_mf_own_bldg_cnt  25695 non-null  float64
 10  very_low_sf_own_bldg_cnt  25695 non-null  float64
 11  low_mf_own_bldg_cnt       25695 non-null  float64
 12  low_sf_own_bldg_cnt       25695 non-null  float64
 13  mod_mf_own_bldg_cnt       25695 non-null  float64
 14  mod_sf

In [17]:
# summing unique entries for each column
df.nunique()

geoid                       25695
gisjoin                     25695
centroid_x                  25695
state_abbr                     44
county_name                  1629
pop_total                    7568
active_subsidies               39
avg_monthly_bill_dlrs        1023
dlrs_kwh                     1023
very_low_mf_own_bldg_cnt    17470
very_low_sf_own_bldg_cnt    25308
low_mf_own_bldg_cnt         15957
low_sf_own_bldg_cnt         25173
mod_mf_own_bldg_cnt         19311
mod_sf_own_bldg_cnt         25346
mid_mf_own_bldg_cnt         20096
mid_sf_own_bldg_cnt         25359
high_mf_own_bldg_cnt        22825
high_sf_own_bldg_cnt        25443
dtype: int64

We have ten owner-occupied building categories (query result columns 7-16). Let's engineer a calculated column to sum the total number of owner occupied buildings in each tract.

In [21]:
# listing the columns to sum (col 7-through-last)
cols_to_sum = df.columns[ 7:]
cols_to_sum

Index(['avg_monthly_bill_dlrs', 'dlrs_kwh', 'very_low_mf_own_bldg_cnt',
       'very_low_sf_own_bldg_cnt', 'low_mf_own_bldg_cnt',
       'low_sf_own_bldg_cnt', 'mod_mf_own_bldg_cnt', 'mod_sf_own_bldg_cnt',
       'mid_mf_own_bldg_cnt', 'mid_sf_own_bldg_cnt', 'high_mf_own_bldg_cnt',
       'high_sf_own_bldg_cnt'],
      dtype='object')

In [22]:
# adding a column to the dataframe ("all_own")
# summing `cols_to_sum` for each row in the dataframe
df['all_own'] = df[cols_to_sum].sum(axis=1)

In [31]:
# reviewing summary statistics
df.describe()

geoid     pop_total  active_subsidies  avg_monthly_bill_dlrs  \
count  2.569500e+04  25695.000000      25695.000000           25695.000000   
mean   3.301156e+10   4483.792683          3.307647             111.248988   
std    1.276324e+10   2091.767106          3.092658              22.411580   
min    1.000104e+10      0.000000          1.000000              46.845787   
25%    2.206250e+10   3031.000000          1.000000              97.396944   
50%    3.403903e+10   4206.000000          2.000000             111.794393   
75%    4.210101e+10   5593.000000          4.000000             127.104540   
max    5.604595e+10  35527.000000         50.000000             262.413841   

           dlrs_kwh  very_low_mf_own_bldg_cnt  very_low_sf_own_bldg_cnt  \
count  25695.000000              25695.000000              25695.000000   
mean       0.128887                  2.252963                 91.658702   
std        0.034146                  3.286216                 60.653672   
min        0.029553                  0.000000                  0.000000   
25%        0.107758                  0.554200                 48.177950   
50%        0.118972                  1.323000                 81.199500   
75%        0.138892                  2.804850                123.687550   
max        0.340253                 73.180300                750.564300   

       low_mf_own_bldg_cnt  low_sf_own_bldg_cnt  mod_mf_own_bldg_cnt  \
count         25695.000000         25695.000000         25695.000000   
mean              1.814383            60.059984             2.930271   
std               2.573076            39.510449             3.680485   
min               0.000000             0.000000             0.000000   
25%               0.287400            32.391250             0.717850   
50%               1.002100            53.732200             1.850800   
75%               2.357900            80.739650             3.880500   
max              56.182500           457.833300           100.469100   

       mod_sf_own_bldg_cnt  mid_mf_own_bldg_cnt  mid_sf_own_bldg_cnt  \
count         25695.000000         25695.000000         25695.000000   
mean            118.653164             3.884647           144.306591   
std              72.110110             5.132888            92.991432   
min               0.000000             0.000000             0.000000   
25%              67.934250             0.834150            78.252900   
50%             111.936400             2.179700           134.853400   
75%             159.741800             4.998850           196.115200   
max             791.294500            99.887400          1363.645500   

       high_mf_own_bldg_cnt  high_sf_own_bldg_cnt  
count          25695.000000          25695.000000  
mean               8.936776            332.245147  
std               13.069552            272.144669  
min                0.000000              0.000000  
25%                1.654650            133.461350  
50%                4.339100            285.136500  
75%               10.673550            462.297350  
max              218.978500           3964.257500

### Subsidies

In [35]:
# finding tracts with the greatest number of subsidies
df[df['active_subsidies']==df['active_subsidies'].max()]

geoid         gisjoin state_abbr      county_name  pop_total  \
20232  36061023000  G3600610023000         NY  New York County       9259   

       active_subsidies  avg_monthly_bill_dlrs  dlrs_kwh  \
20232              50.0              98.685298  0.249082   

       very_low_mf_own_bldg_cnt  very_low_sf_own_bldg_cnt  \
20232                    1.2661                    3.0588   

       low_mf_own_bldg_cnt  low_sf_own_bldg_cnt  mod_mf_own_bldg_cnt  \
20232               0.7864               1.1471               1.2572   

       mod_sf_own_bldg_cnt  mid_mf_own_bldg_cnt  mid_sf_own_bldg_cnt  \
20232               2.2941               1.2882               1.1471   

       high_mf_own_bldg_cnt  high_sf_own_bldg_cnt  
20232                1.7296                1.5294

In [58]:
# subset columns from tracts with the lowest number of subsidies
df.loc[df['active_subsidies']==
       df['active_subsidies'].min(), 
       ['geoid', 'state_abbr', 'county_name', 'very_low_mf_own_bldg_cnt']]

geoid state_abbr        county_name  very_low_mf_own_bldg_cnt
0      41017000500         OR   Deschutes County                    0.2681
5      28029950500         MS      Copiah County                    0.0000
6      28033070323         MS      DeSoto County                    0.9087
7      28033071110         MS      DeSoto County                    0.8272
13     28043950200         MS     Grenada County                    0.0000
...            ...        ...                ...                       ...
25679  12087971200         FL      Monroe County                    2.9583
25688  12086003915         FL  Miami-Dade County                    5.4488
25690  45019002802         SC  Charleston County                    0.6422
25692  34009021300         NJ    Cape May County                    5.3800
25693  34023008900         NJ   Middlesex County                    3.2189

[8088 rows x 4 columns]

### Solar suitable buildings per capita

Before we can calculate the number of qualified buildings per capita, we need to better understand a few of our columns. 

The data documentation indicates that the `...bldg_cnt` columns represent building counts. However, as we inspect the data we observe that the values are floats.

What does it mean that Oregon's Deschutes County has a 'Very Low Income (0-30% AMI) - Single-Family -
Owner-Occupied - Household Count' of 0.2681?

======================================================
======================================================

===== code for dropping the table (e.g., to avoid duplicating records if re-running the notebook) =====

In [59]:
# # creating a connection to the database
# conn = sqlite3.connect("data/rep.sqlite")

# # instantiating a cursor
# cur = conn.cursor()

# # dropping the table
# cur.execute("""DROP TABLE seed
#             """);